### PR TITLE
Detect flaky performance test scenarios

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

As part of https://github.com/gradle/gradle-private/issues/1635 , we want to detect flaky performance test with a weekly job, in order to know which scenarios are flaky. 

This PR does:

- Add a `repeat` property to `DistributedPerformanceTest` to control how many times a performance coordinator is executed. For `FlakinessDetection` this number is 3 - all cross version scenarios are repeated 3 times.
- Refactor `ReportGenerator` class hierarchy to support slightly different flakiness report generation:

```
AbstractReportGenerator
├── DefaultReportGenerator
├── BuildScanReportGenerator
└── FlakinessReportGenerator
```
- Refactor `IndexPageGenerator` to reuse its code when generating flakiness report. The result report looks like this: https://builds.gradle.org/viewLog.html?buildId=19553872&buildTypeId=Gradle_Check_PerformanceTestCoordinator&tab=report_project944_Performance&branch_Gradle_Check_Stage_ReadyforMerge=blindpirate%2Fdetect-performance-flakiness

### CODE REVIEW

File: .teamcity/Gradle_Check/model/CIBuildModel.kt [hunk 1]
```diff
@@ -284,8 +284,9 @@ enum class JvmVendor {
 }
 
 enum class PerformanceTestType(val taskId: String, val timeout : Int, val defaultBaselines: String = "", val extraParameters : String = "") {
-    test("PerformanceTest", 420, "defaults"),
+    test("PerformanceTest", 420, "nightly"),
```
Comment (jjohannes) [inline .teamcity/Gradle_Check/model/CIBuildModel.kt]: Is this change on purpose?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/AbstractReportGenerator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        FileRenderer fileRenderer = new FileRenderer();
+        TestPageGenerator testHtmlRenderer = new TestPageGenerator(projectName);
+        TestDataGenerator testDataRenderer = new TestDataGenerator();
+        fileRenderer.render(testResults, testHtmlRenderer, new File(outputDirectory, "tests/" + testResults.getId() + ".html"));
+        fileRenderer.render(testResults, testDataRenderer, new File(outputDirectory, "tests/" + testResults.getId() + ".json"));
+    }
+
+    protected ResultsStore getResultsStore() throws Exception {
+        Type superClass = getClass().getGenericSuperclass();
+        Class<? extends ResultsStore> resultsStoreClass = (Class<T>) ((ParameterizedType) superClass).getActualTypeArguments()[0];
```
Comment (wolfs) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/AbstractReportGenerator.java]: You can use Guava's `TypeToken` here:
Something like:
```
ParameterizedType type = (ParameterizedType) beanType.getSupertype(AbstractReportGenerator).getType();
TypeToken.of(type.getActualTypeArguments()[typeParameterIndex]);
```

File: .teamcity/Gradle_Check/model/CIBuildModel.kt [hunk 3]
```diff
@@ -284,8 +284,9 @@ enum class JvmVendor {
...
     experiment("PerformanceExperiment", 420, "defaults"),
+    flakinessDetection("FlakinessDetection", 420, "nightly"),
```
Comment (wolfs) [inline .teamcity/Gradle_Check/model/CIBuildModel.kt]: So we are running flakiness detection against the latest nightly now, which automatically turns of caching?

Would it be much harder to use the current commit as a baseline?

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/ScenarioBuildResultData.groovy [hunk 4]
```diff
@@ -90,6 +91,10 @@ class ScenarioBuildResultData {
         }
     }
 
+    boolean isFlaky() {
+        return executions.any { it.confidencePercentage > 95 }
```
[the remark this answers is not in the document]
Comment (jjohannes) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/ScenarioBuildResultData.groovy]: Oh I missed this. I agree that we should start with a higher threshold, but I don't have a good feeling for what a good threshold is. Maybe 99% is good to start with if you agree Bo.

File: .teamcity/Gradle_Check/model/CIBuildModel.kt (hunk 3 above)
Comment (blindpirate) [reply .teamcity/Gradle_Check/model/CIBuildModel.kt]: Now I introduced a new baseline `flakiness-detection-commit` and build cache is disabled upon this baseline. It will be resolved to current commit.

**Comment by blindpirate** [comment]
@wolfs @jjohannes PTAL. If it looks fine I'll try it tomorrow as weekly job.

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/DetermineBaselines.kt [hunk 5]
```diff
@@ -53,6 +58,9 @@ open class DetermineBaselines : DefaultTask() {
     private
     fun Property<String>.isDefaultValue() = !isPresent || get() in listOf("", defaultBaseline, Config.baseLineList)
 
+    private
+    fun currentCommitBaseline() = commitBaseline(project.execAndGetStdout("git", "rev-parse", "HEAD"))
```
Comment (wolfs) [inline buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/DetermineBaselines.kt]: You could make this a `val` with `get`:
```
val currentCommitBaseline
    get() = commitBaseline(project.execAndGetStdout("git", "rev-parse", "HEAD"))
```

Then you could reference it as a value instead of a method.
